tools/build/third_party/breakpad/dump_syms_pe.cc

```diff
@@ -0,0 +1,389 @@
+  uint32_t timeStamp;
+  uint32_t symbolTableOffset;
+  uint32_t numSymbols;
+  uint16_t extraHeaderSize;
+  uint16_t flags;
+} PEFileHeader;
+
+typedef struct PEOptionalHeader {
+  uint16_t signature;
+  uint16_t linkerVersion;
+  uint32_t sizeOfCode;
+  uint32_t sizeOfInitData;
+  uint32_t sizeOfUnintData;
+  uint32_t entryPoint;
+  uint32_t codeBase;
+  uint64_t imageBase;
+  uint32_t sectionAlign;
+  uint32_t fileAlign;
+  uint32_t osVersion;
+  uint32_t imageVersion;
+  uint32_t subsystemVersion;
+  uint32_t winVersion;
+  uint32_t imageSize;
+  uint32_t headerSize;
+  uint32_t checksum;
+  uint16_t subsystem;
+  uint16_t dllFlags;
+  uint64_t stackReserve;
+  uint64_t stackCommit;
+  uint64_t heapReserve;
+  uint64_t heapCommit;
+  uint32_t loaderFlags;
+  uint32_t numDirEntries;
+  struct dirEntry {
+    uint32_t address;
+    uint32_t size;
+  } directory[7]; // We only care about the debug entry, which is the 7th.
+} PEOptionalHeader;
+
+typedef struct PESectionHeader {
+  char name[8];
+  uint32_t virtualSize;
+  uint32_t virtualAddress;
+  uint32_t size;
+  uint32_t offset;
+  uint32_t reloctionsOffset;
+  uint32_t lineNumbersOffset;
+  uint16_t numRelocations;
+  uint16_t numLineNumbers;
+  uint32_t flags;
+} PESectionHeader;
+
+#define SYMBOL_SIZE 18
+
+typedef struct StringTable {
+  const char* data;
+  int size;
+} StringTable;
+
+typedef struct PEFile {
+  std::vector<char> data;
+  PEFileHeader* header;
+  PEOptionalHeader* optionalHeader;
+  PESectionHeader* sections;
+  StringTable strings;
+  std::string debugId;
+  std::string codeId;
+} PEFile;
+
+typedef struct DebugDirectory {
+  uint32_t signature;
+  uint32_t timestamp;
+  uint32_t version;
+  uint32_t type;
+  uint32_t size;
+  uint32_t address;
+  uint32_t offset;
+} DebugDirectory;
+
+typedef struct DebugData {
+  uint32_t signature;
+  struct {
+    uint32_t a;
+    uint16_t b;
+    uint16_t c;
+    uint8_t d[8];
+  } uuid;
+  uint32_t age;
+} DebugData;
+
+static int getBase(const std::vector<char>& data) {
+  int base = 0;
+  if (data.size() >= 64 && data[0] == 'M' && data[1] == 'Z') {
+    int offset = ((uint32_t *)data.data())[15];
+    if (offset + 4 < data.size() && data[offset] == 'P' && data[offset + 1] == 'E' && data[offset + 2] == 0 && data[offset + 3] == 0) {
+      base = offset + 4;
+    }
+  }
+  return base;
+}
+
+static PESectionHeader* findSection(const PEFile& file, uint32_t address) {
+  for (int i = 0; i < file.header->numSections; i++) {
+    PESectionHeader* section = &file.sections[i];
+    if (section->virtualAddress <= address && section->virtualAddress + section->virtualSize > address) {
+      return section;
+    }
+  }
+  return nullptr;
+}
+
+// buffer needs to be 9 bytes (or larger).
+static const char* getSectionName(const PEFile& file, const PESectionHeader& section, char* buffer) {
+  if (section.name[0] == '/') {
+    char offS[8];
+    std::memcpy(offS, &section.name[1], 7);
+    offS[7] = 0;
+    int off = atoi(offS);
```

Comment:
`int off = std::stoi(std::string(section.name, 1, 7));`

Note that an exception will be thrown if the number cannot be parsed.